### PR TITLE
Make log of rules triggered to be per query in iterative optimizer

### PR DESCRIPTION
**Pull request by feilong-liu**
Sample test failure https://github.com/prestodb/presto/actions/runs/3452381058/jobs/5762250996

Optimizers are singleton in coordinator shared with all queries, make the logging part to be per query by adding logging code to context in iterative optimizer.

Also make access to optimizer information collector to be synchronized (although it should not be a problem as it's a new instance for each session), but still make it safe here just in case.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerInformationCollector.java
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerInformationCollector.java:27]: Hmm is this better or do we use Atomic

**Review by kaikalur — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerInformationCollector.java
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerInformationCollector.java:27]: this doesn't actually make it synchronized, it's just an annotation, you would need to also make all the access methods synchronized.  Alternatively, you can use a ConcurrentLinkedQueue.  In either case, I would make an immutableCopy before returning in getOptimizationInfo().  otherwise the caller can modify the object.

**Comment by rschlussel** [comment]
Also, I would recommend reading Java Concurrency in Practice as a great reference for dealing with concurrency in Java.

**Review by rschlussel — APPROVED** [review]

**Review by rschlussel — CHANGES_REQUESTED** [review]
You still have a ConcurrentModificationException.  Looks like maybe `rulesTriggered` gets modified as you iterate through. 
```
2022-11-14T06:29:47.5150442Z [ERROR] com.facebook.presto.tpcds.TestTpcds.testLargeInWithShortDecimal  Time elapsed: 0.983 s  <<< FAILURE!
2022-11-14T06:29:47.5161219Z java.lang.AssertionError: Execution of 'actual' query failed: SELECT typeof(i_current_price) FROM item LIMIT 1
2022-11-14T06:29:47.5161721Z 	at org.testng.Assert.fail(Assert.java:98)
2022-11-14T06:29:47.5162234Z 	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:178)
2022-11-14T06:29:47.5162851Z 	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:106)
2022-11-14T06:29:47.5165730Z 	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:154)
2022-11-14T06:29:47.5166894Z 	at com.facebook.presto.tpcds.TestTpcds.testLargeInWithShortDecimal(TestTpcds.java:70)
2022-11-14T06:29:47.5168261Z 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2022-11-14T06:29:47.5169260Z 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2022-11-14T06:29:47.5170624Z 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2022-11-14T06:29:47.5171372Z 	at java.lang.reflect.Method.invoke(Method.java:498)
2022-11-14T06:29:47.5173549Z 	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:135)
2022-11-14T06:29:47.5174469Z 	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:673)
2022-11-14T06:29:47.5176913Z 	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:220)
2022-11-14T06:29:47.5177990Z 	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
2022-11-14T06:29:47.5180706Z 	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:945)
2022-11-14T06:29:47.5181580Z 	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:193)
2022-11-14T06:29:47.5184050Z 	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
2022-11-14T06:29:47.5185083Z 	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
2022-11-14T06:29:47.5186027Z 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
2022-11-14T06:29:47.5188452Z 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
2022-11-14T06:29:47.5189229Z 	at java.lang.Thread.run(Thread.java:750)
2022-11-14T06:29:47.5190061Z Caused by: java.lang.RuntimeException: java.util.ConcurrentModificationException
2022-11-14T06:29:47.5191078Z 	at com.facebook.presto.tests.AbstractTestingPrestoClient.execute(AbstractTestingPrestoClient.java:124)
2022-11-14T06:29:47.5192194Z 	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:726)
2022-11-14T06:29:47.5193204Z 	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:694)
2022-11-14T06:29:47.5194227Z 	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:175)
2022-11-14T06:29:47.5194994Z 	... 18 more
2022-11-14T06:29:47.5195563Z Caused by: java.util.ConcurrentModificationException
2022-11-14T06:29:47.5196385Z 	at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1585)
2022-11-14T06:29:47.5197194Z 	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:647)
2022-11-14T06:29:47.5198272Z 	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.optimize(IterativeOptimizer.java:132)
2022-11-14T06:29:47.5199337Z 	at com.facebook.presto.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:196)
2022-11-14T06:29:47.5200113Z 	at com.facebook.presto.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:184)
2022-11-14T06:29:47.5201154Z 	at com.facebook.presto.execution.SqlQueryExecution.lambda$createLogicalPlan$2(SqlQueryExecution.java:477)
2022-11-14T06:29:47.5204556Z 	at com.facebook.presto.common.RuntimeStats.profileNanos(RuntimeStats.java:136)
2022-11-14T06:29:47.5206899Z 	at com.facebook.presto.execution.SqlQueryExecution.createLogicalPlan(SqlQueryExecution.java:475)
2022-11-14T06:29:47.5207791Z 	at com.facebook.presto.execution.SqlQueryExecution.start(SqlQueryExecution.java:405)
2022-11-14T06:29:47.5209222Z 	at com.facebook.presto.$gen.Presto_0_279_SNAPSHOT_a634b63__testversion____20221114_062912_4.run(Unknown Source)
2022-11-14T06:29:47.5210442Z 	at com.facebook.presto.execution.SqlQueryManager.createQuery(SqlQueryManager.java:286)
2022-11-14T06:29:47.5211904Z 	at com.facebook.presto.dispatcher.LocalDispatchQuery.lambda$startExecution$8(LocalDispatchQuery.java:211)
2022-11-14T06:29:47.5212581Z 	... 3 more
```

**Comment by kaikalur** [comment]
> the synchronized changes are good, but don't fix the problem. I would move rulesTriggered to the Context inner class. I think this issue is coming from having multiple queries running at a time, and their rulesTriggered are all getting added to the same set. You only want the rules to last for the lifetime of a single query. Current implementation will actually be a big problem for a long running cluster with lots of queries.
> 
> I would also consider making the OptimizerInformationCollector have a set of PlanOptimizerInformation instead of a list, and then pass the OptimizerInformationCollector directly to the context of tracking rulesTriggered.

Hmm - that's confusing. Isn't a session unique to a query? In fact, I would think this synchronization should not needed at all - after looking at the change.

**Comment by rschlussel** [comment]
yes, it's not likely the class will be accessed from multiple threads (though theoretically possible, so better to be safe).  The problem is with rulesTriggered, which is global to the IterativeOptimizer instance.  I think Optimizers are singletons shared between queries, so they should not be carrying state.  That's where you get the concurrent modifications (and also incorrect stats).  Any state specific to a particular query should go in the Context

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:132]: the concurrent modification problem is with rulesTriggered, not the OptimizationInformationCollector.

**Comment by feilong-liu** [comment]
> I think Optimizers are singletons shared between queries

Indeed it's a singleton in coordinator. Thanks for the information
https://github.com/prestodb/presto/blob/c78a75a4983fdd0dc265e5cd3b2d2da5a9bdaf89/presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java#L243

> I would also consider making the OptimizerInformationCollector have a set of PlanOptimizerInformation instead of a list

Changed the list to a set.
This means that if two rules are triggered in multiple iterative optimizers, we are to only record one of it. Since most of rules which gets triggered multiple times are plan simplification rules, and not of much interest. I guess we can leave it as it is.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerInformationCollector.java
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerInformationCollector.java:29]: > Hmm is this better or do we use Atomic

Since the OptimizerInformationCollector is per session object, we should not hit concurrency issue. Use synchronized method to be consistent to warning collector here.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerInformationCollector.java
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerInformationCollector.java:37]: return a set

**Review by rschlussel — COMMENTED** [review]
thanks, this is good

**Comment by kaikalur** [comment]
> > I think Optimizers are singletons shared between queries
> 
> Indeed it's a singleton in coordinator (although this seems a bit suprising and counter intuitive to me). Thanks for the information
> 
> https://github.com/prestodb/presto/blob/c78a75a4983fdd0dc265e5cd3b2d2da5a9bdaf89/presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java#L243
> 
> > I would also consider making the OptimizerInformationCollector have a set of PlanOptimizerInformation instead of a list
> 
> Changed the list to a set. This means that if two rules are triggered in multiple iterative optimizers, we are to only record one of it. Since most of rules which gets triggered multiple times are plan simplification rules, and not of much interest. I guess we can leave it as it is.

It could still be useful to list them separately and in the future adding more information like which part of the query triggered a particular instance

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerInformationCollector.java
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerInformationCollector.java:37]: I was intentionally using an LInkedHashSet hoping to preserve the order of which the rules are triggered. If we return a 
copy of Immutable set here, will the order be lost?

Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerInformationCollector.java:37]: I think immutable set maintains insertion order https://guava.dev/releases/19.0/api/docs/com/google/common/collect/ImmutableSet.html#copyOf(java.util.Collection)

**Comment by rschlussel** [comment]
> > > I think Optimizers are singletons shared between queries
> > 
> > 
> > Indeed it's a singleton in coordinator (although this seems a bit suprising and counter intuitive to me). Thanks for the information
> > https://github.com/prestodb/presto/blob/c78a75a4983fdd0dc265e5cd3b2d2da5a9bdaf89/presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java#L243
> > 
> > > I would also consider making the OptimizerInformationCollector have a set of PlanOptimizerInformation instead of a list
> > 
> > 
> > Changed the list to a set. This means that if two rules are triggered in multiple iterative optimizers, we are to only record one of it. Since most of rules which gets triggered multiple times are plan simplification rules, and not of much interest. I guess we can leave it as it is.
> 
> It could still be useful to list them separately and in the future adding more information like which part of the query triggered a particular instance

I don't feel strongly about this

**Comment by feilong-liu** [comment]
Keep returning a list of optimization information rather than a set here.

**Review by rschlussel — APPROVED** [review]

**Comment by rschlussel** [comment]
@kaikalur can you review and approve if you have no more comments?  The changes requested blocks merging.

**Review by kaikalur — APPROVED** [review]
LGTM

**Pull request merged**
